src/main/java/com/bobocode/hoverla/bring/context/BeanFactory.java

```diff
@@ -0,0 +1,61 @@
+package com.bobocode.hoverla.bring.context;
+
+import com.bobocode.hoverla.bring.exception.NoSuchBeanException;
+import com.bobocode.hoverla.bring.exception.NoUniqueBeanException;
+
+import java.util.Map;
+
+/**
+ * Convenient interface to work with IoC container, which provides read API only.
+ * Allows getting registered beans by their names and types.
+ */
+public interface BeanFactory {
+
+    /**
+     * Returns a bean instance by its type.
+     *
+     * @param beanType type of the desired instance
+     * @param <T>      type of the desired instance
+     * @return object of type T which represents desired instance
+     * @throws NoSuchBeanException   if instance with provided type is not found
+     * @throws NoUniqueBeanException if more than one instance with provided type is found
+     */
+    <T> T getBean(Class<T> beanType);
+
+    /**
+     * Returns a bean instance by its name.
+     *
+     * @param beanName name of the desired instance
+     * @return instance object
+     * @throws NoSuchBeanException if instance with provided name and type is not found
+     */
+    Object getBean(String beanName);
+
+    /**
+     * Returns a bean instance by its name and type.
+     *
+     * @param beanName name of the desired instance
+     * @param beanType type of the desired instance
+     * @param <T>      type of the desired instance
+     * @return object of type T which represents desired instance
+     * @throws NoSuchBeanException if instance with provided name and type is not found
+     */
+    <T> T getBean(String beanName, Class<T> beanType);
```

Comment:
May you explain please, why is this method needed? All beanNames are unique as I understand, so beanType here adds no qualifying

Reply:
@nourish1709 What if we have 3 different beans of the same type (interface for example) ? Beans are unique by name, not by type.

Reply:
> @nourish1709 What if we have 3 different beans of the same type (interface for example) ? Beans are unique by name, not by type.

as I understand, bean type just adds type safety, doesn't it?

Reply:
@nourish1709 exactly. It is similar to `getBean(String name)`, but it is type safe.